ClosureCompilerRecursive.java

```diff
@@ -0,0 +1,123 @@
+import com.google.javascript.jscomp.Compiler;
+import com.google.javascript.jscomp.CompilerOptions;
+import com.google.javascript.jscomp.SourceFile;
+import com.google.javascript.jscomp.CommandLineRunner;
+import com.google.javascript.jscomp.CompilerOptions.LanguageMode;
+
+public class ClosureCompilerRecursive {
+   protected Path directory;
+   protected int errorCount = 0;
+   protected boolean writeFiles = true;
+
+   protected String compile(File file) {
+      Compiler compiler = new Compiler();
+
+      List<SourceFile> list = null;
+
+      try {
+          list =
+         CommandLineRunner
+         .getBuiltinExterns(CompilerOptions.Environment.BROWSER);
+      } catch (IOException e) {
+          System.out.println(e.getMessage());
+          System.exit(1);
+      }
+
+      list.add(SourceFile.fromFile(file.getPath()));
+      compiler.compile(new ArrayList<SourceFile>(), list, getOptions());
+      errorCount += compiler.getResult().errors.length;
+      return compiler.toSource();
+   }
+
+   public static void main(String[] args) throws IOException {
+      ClosureCompilerRecursive ccr = new ClosureCompilerRecursive();
+      setOptionsFromArgs(ccr, args);
+      ccr.compileDirectory();
+      System.exit(Math.min(ccr.errorCount, 127));
+   }
+
+   protected static void setOptionsFromArgs(ClosureCompilerRecursive ccr, String[] args) {
+      String dir = null;
+      if (args.length == 1) {
+         dir = args[0];
+      } else if (args.length == 2 && args[0].equals("--validate-syntax")) {
+         ccr.validateSyntaxOnly();
+         dir = args[1];
+      } else {
+         printHelp();
+      }
+
+      ccr.setDirectory(Paths.get(dir));
+   }
+
+   protected static void printHelp() {
+      System.out.println("Usage: closure-compiler-recursive [--validate-syntax] /path/to/directory");
+      System.exit(1);
+   }
+
+   public void compileDirectory() throws IOException {
+      for(File file : getFilesFromDirectory()) {
+         String compressedSource = compile(file);
+         if (writeFiles) {
+            writeFile(file, compressedSource);
+         }
+         System.gc();
+      }
+   }
+
+   public void validateSyntaxOnly() {
+      writeFiles = false;
+   }
+
+   public void setDirectory(Path dir) {
+      directory = dir;
+   }
+
+   protected void writeFile(File file, String contents) throws IOException {
+      PrintWriter writer = new PrintWriter(new BufferedWriter(new FileWriter(file.toString())));
+      writer.println(contents);
+      writer.close();
+   }
+
+   protected File[] getFilesFromDirectory() throws IOException {
+      File[] files = Files.find(
+         directory,
+         Integer.MAX_VALUE,
+           (path, fileAttr) -> fileAttr.isRegularFile()
+                               && path.toString().endsWith(".js"))
+         .map((Path path) -> path.toFile())
+         .toArray(File[]::new);
+      Arrays.sort(files, (File f1, File f2) -> (int) (f2.length() - f1.length()));
```

Comment:
Why sort the files by length?

Reply:
Think this should be in a comment?
a22bf7b7fc6a3bc8b0b8fa11d9405c31d71f886a

Reply:
I don't know :/ Does the comment's helpfulness outweight it's inherent evilness? 😂